doc/rtd/howto/return_codes.rst

```diff
@@ -0,0 +1,172 @@
+the instance, or if cloud-init internally experienced an error, this status
+is now visible via :code:`cloud-init status --long` after the `extended_status`
+key. Additionally, the cause of this error will be visible after the
+`recoverable_errors` key.
+
+Return codes
+""""""""""""
+
+Cloud-init historically used two return codes from the `cloud-init status`
+subcommand: 0 to indicate success and 1 to indicate failure. These return codes
+lacked nuance. Return code 0 (success) included the in-between when something
+went wrong, but cloud-init was able to finish.
+
+Many users of cloud-init run :code:`cloud-init status --wait` and expect that
+when complete, cloud-init has finished. Since cloud-init is not guaranteed to
+succeed, users should also be check the return code of this command.
+
+A backwards-incompatible change was recently introduced: errors that did not
+crash cloud-init now have an exit code of 2. After this change, exit 1 still
+means that cloud-init crashed, but exit code 0 now more correctly means that
+cloud-init succeeded. Anyone that previously checked for exit code 0 should
+probably update their assumptions in one of the following two ways:
+
+Users that wish to take advantage of cloud-init's new error reporting
+capabilities should check for exit code of 2 from :code:`cloud-init status`.
+An example of this:
+
+.. code-block:: python
+
+    from logging import getLogger
+    from json import loads
+    from subprocess import run
+    from sys import exit
+
+    logger = getLogger(__name__)
+    completed = run("cloud-init status --format json")
+    output = loads(completed.stdout)
+
+    if 2 == completed.return_code:
+        # something bad might have happened - we should check it out
+        logger.warning("cloud-init experienced a recoverable error")
+        logger.warning("status: %s", output.get("extended_status"))
+        logger.warning("recoverable error: %s", output.get("recoverable_errors"))
+
+    elif 1 == completed.return_code:
+        # cloud-init completely failed
+        logger.error("cloud-init crashed, all bets are off!")
+        exit(1)
+
+Users that wish to use ignore cloud-init's errors and check the return code in
+a backwards-compatible way should check that the return code is not equal to
+1. This will provide the same behavior before and after the changed exit code.
+See an example of this:
+
+.. code-block:: python
+
+    from logging import getLogger
+    from subprocess import run
+    from sys import exit
+
+    logger = getLogger(__name__)
+    completed = run("cloud-init status --format json")
+
+    if 1 == completed.return_code:
+        # cloud-init completely failed
+        logger.error("cloud-init crashed, all bets are off!")
+        exit(1)
+
+    # cloud-init might have failed, but this code ignores that possibility
+    # in preference of backwards compatibility
+
+Some distros (such as RHEL and Ubuntu) are likely to patch out this
```

Comment:
Not sure I'd call out specific distros here.

Also, this paragraph feels a little weird to call out here. We don't generally point out every possible breaking change that has been patched out in various downstreams. Not sure we need anything, but if so I'd reduce it to a note with something along the lines of "Distributions have likely removed this behavior for stable releases".